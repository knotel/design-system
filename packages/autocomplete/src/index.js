import React from 'react'
import PropTypes from 'prop-types'
import Downshift from 'downshift'
import styled from 'styled-components'
import {
  Flex,
  Card,
  Label as PclnLabel,
  Input as PclnInput
} from 'pcln-design-system'
import { themeGet } from 'styled-system'

export const AutocompleteContext = React.createContext()

export const withAutocomplete = (Component, mapProps) =>
  React.forwardRef((props, ref) => (
    <AutocompleteContext.Consumer
      children={state => (
        <Component ref={ref} {...props} {...mapProps({ ...props, ...state })} />
      )}
    />
  ))

export const Label = withAutocomplete(PclnLabel, ({ getLabelProps }) =>
  getLabelProps()
)
Label.isLabel = true

export const Input = withAutocomplete(PclnInput, ({ getInputProps }) =>
  getInputProps()
)
Input.isField = true

const MenuCard = styled(Card)`
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  max-height: ${props => props.height};
  overflow-y: auto;
`

MenuCard.defaultProps = {
  bg: 'white',
  height: '256px',
  borderWidth: 0,
  boxShadowSize: 'lg',
  mt: 1
}

const MenuRoot = React.forwardRef((props, ref) => (
  <MenuCard {...props} innerRef={ref} />
))

export const Menu = ({ children, ...props }) => (
  <AutocompleteContext.Consumer
    children={({ match, isOpen, getMenuProps, inputValue }) => (
      <MenuRoot
        {...getMenuProps({
          ...props,
          isOpen
        })}
        children={
          isOpen
            ? React.Children.toArray(children)
                .filter(el => match(el.props.item, inputValue))
                .map((el, index) => React.cloneElement(el, { index }))
            : false
        }
      />
    )}
  />
)

const ItemRoot = styled(Flex)`
  cursor: pointer;
  &[aria-selected='true'] {
    background-color: ${themeGet('colors.lightBlue')};
  }
  &[data-highlighted] {
    color: ${themeGet('colors.white')};
    background-color: ${themeGet('colors.blue')};
    & svg {
      color: ${themeGet('colors.white')};
    }
  }
`

ItemRoot.defaultProps = {
  alignItems: 'center'
}

export const Item = props => (
  <AutocompleteContext.Consumer
    children={({ item, getItemProps, highlightedIndex }) => (
      <ItemRoot
        {...getItemProps({
          item,
          px: 2,
          py: 1,
          'data-highlighted':
            highlightedIndex === props.index ? true : undefined,
          ...props
        })}
      />
    )}
  />
)

export class Autocomplete extends React.Component {
  static Label = Label
  static Input = Input
  static Menu = Menu
  static Item = Item

  static propTypes = {
    match: PropTypes.func
  }

  static defaultProps = {
    match: () => true
  }

  render() {
    const { children, style, ...props } = this.props

    return (
      <Downshift
        {...props}
        children={state => (
          <div style={{ position: 'relative', ...style }}>
            <AutocompleteContext.Provider
              value={{ ...props, ...state }}
              children={children}
            />
          </div>
        )}
      />
    )
  }
}
