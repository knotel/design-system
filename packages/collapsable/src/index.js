// import React, { Component } from 'react'
import Collapsable from 'react-collapsible'

export default Collapsable
