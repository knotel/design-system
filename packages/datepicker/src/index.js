import styled from 'styled-components'
import {
  DateRangePicker,
  SingleDatePicker,
  DayPickerRangeController
} from 'react-dates'
import { space, color, themeGet } from 'styled-system'

export const Datepicker = styled(DateRangePicker)`
  ${space} ${color} 

  .CalendarDay__selected_span {
    background: #82e0aa; //background
    color: white; //text
    border: 1px solid ${themeGet('colors.alphablue')}; //default styles include a border
  }

  // Will edit selected date or the endpoints of a range of dates
  .CalendarDay__selected {
    background: ${themeGet('colors.alphablue')};
    color: white;
  }

  // Will edit when hovered over. _span style also has this property
  .CalendarDay__selected:hover {
    background: ${themeGet('colors.alphablue')};
    color: white;
  }

  .CalendarDay__hovered_span:hover,
  .CalendarDay__hovered_span {
    background: ${themeGet('colors.alphablue')};
  }
`

Datepicker.defaultProps = {
  color: 'blue'
}

Datepicker.propTypes = {
  ...space.propTypes,
  ...color.propTypes
}

export default Datepicker
