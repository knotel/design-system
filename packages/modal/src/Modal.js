import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Transition } from 'react-transition-group'
import { color, width, height } from 'styled-system'
import { DialogOverlay, DialogContent } from '@reach/dialog'
import { Box, CloseButton, Flex } from 'pcln-design-system'

const Overlay = styled(DialogOverlay)`
  background-color: rgba(0, 0, 0, 0.7);
  bottom: 0;
  left: 0;
  overflow-y: auto;
  position: fixed;
  right: 0;
  top: 0;
  z-index: ${props => props.zindex || 100};
  font-family: ${props => props.theme.font};
  opacity: 0;
  transition: opacity .5s cubic-bezier(0.50, 0.00, 0.25, 1.00);

  ${props =>
    props.transitionstate === 'entering' &&
    `
    opacity: 0;
  `}

  ${props =>
    props.transitionstate === 'entered' &&
    `
    opacity: 1;
  `}

  ${props =>
    props.transitionstate === 'exiting' &&
    `
    opacity: 0;
  `}

  ${props =>
    props.transitionstate === 'exited' &&
    `
    opacity: 0;
  `}
`

const Dialog = styled(DialogContent)`
  ${color} 
  ${width}
  ${height}
  max-width: calc(100vw - 32px);
  position: relative;
  margin-left: auto;
  margin-right: auto;
  box-shadow: ${props => props.theme.boxShadows[3]};
  &:focus {
    outline: none;
  }
  max-height: 100vh;
  ${props =>
    props.enableoverflow &&
    `
    max-height: none;
  `}

  transform: scale(0.5);
  transition: transform .5s cubic-bezier(0.50, 0.00, 0.25, 1.00);

  ${props =>
    props.transitionstate === 'entering' &&
    `
    transform: scale(0.5);
  `}

  ${props =>
    props.transitionstate === 'entered' &&
    `
    transform: scale(1);
  `}

  ${props =>
    props.transitionstate === 'exiting' &&
    `
    transform: scale(0.5);
  `}

  ${props =>
    props.transitionstate === 'exited' &&
    `
    transform: scale(0.5);
  `}
`

const HeaderWrapper = styled.div`
  flex-shrink: 0;
`

const FloatCloseButton = styled(CloseButton)`
  position: absolute;
  top: 16px;
  right: 16px;
  z-index: 2;

  svg {
    vertical-align: top;
  }

  &:focus {
    background-color: ${props => props.theme.colors.borderGray};
    outline: none;
  }
`

const ContentWrapper = styled(Box)`
  position: relative;
  ${props => {
    if (!props.enableoverflow) {
      return `
        overflow-y: scroll;
        overflow-x: hidden;
        height: 100%;
        -webkit-overflow-scrolling: touch;
      `
    }
  }}
`

const OverlayWrapper = styled.div`
  display: table;
  height: 100%;
  width: 100%;
`

const DialogWrapper = styled(Box)`
  display: table-cell;
  vertical-align: middle;
  ${props =>
    props.enableoverflow &&
    `
    padding-top: 24px;
    padding-bottom: 24px;
  `};
`

const DialogInnerWrapper = styled(Flex)`
  position: relative;
  height: 100%;
`

const Modal = ({
  isOpen,
  onClose,
  bg,
  zIndex,
  children,
  imgMode,
  width,
  className,
  header,
  disableCloseButton,
  enableOverflow,
  height
}) => (
  <Transition in={isOpen} unmountOnExit timeout={500}>
    {state => (
      <Overlay
        onDismiss={onClose}
        zindex={zIndex}
        transitionstate={state}
        initialFocusRef={null}
      >
        <OverlayWrapper>
          <DialogWrapper enableoverflow={enableOverflow}>
            <Dialog
              width={width}
              bg={bg}
              height={enableOverflow ? null : height}
              transitionstate={state}
              className={className}
              enableoverflow={enableOverflow ? 'true' : null}
            >
              <DialogInnerWrapper flexDirection="column">
                {header && <HeaderWrapper>{header}</HeaderWrapper>}
                <ContentWrapper
                  p={imgMode ? 0 : 16}
                  header={header}
                  enableoverflow={enableOverflow}
                >
                  {enableOverflow && !disableCloseButton && (
                    <FloatCloseButton
                      data-testid="pcln-modal-close"
                      header={header}
                      onClick={onClose}
                      tabIndex="1"
                    />
                  )}
                  {children}
                </ContentWrapper>
              </DialogInnerWrapper>
            </Dialog>
          </DialogWrapper>
        </OverlayWrapper>
      </Overlay>
    )}
  </Transition>
)

Modal.defaultProps = {
  isOpen: false,
  disableCloseButton: false,
  enableOverflow: false,
  header: null,
  bg: 'white',
  height: 420
}

Modal.propTypes = {
  ...width.propTypes,
  ...height.propTypes,
  isOpen: PropTypes.bool,
  disableCloseButton: PropTypes.bool,
  enableOverflow: PropTypes.bool,
  onClose: PropTypes.func,
  bg: PropTypes.string,
  zIndex: PropTypes.number,
  title: PropTypes.string,
  headerBg: PropTypes.string,
  imgMode: PropTypes.bool,
  className: PropTypes.string,
  header: PropTypes.any
}

export default Modal
