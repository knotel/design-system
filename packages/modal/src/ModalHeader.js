import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Flex, CloseButton, Text } from 'pcln-design-system'

const HeaderWrapper = styled(Flex)`
  height: 40px;
`

const StyledCloseButton = styled(CloseButton)`
  svg {
    vertical-align: top;
  }

  &:focus {
    background-color: ${props => props.theme.colors.darkBlue};
    outline: none;
  }
`

const ModalHeader = ({ bg, color, onClose, title }) => (
  <HeaderWrapper align="center" color={color} bg={bg} px={3}>
    {title && (
      <Text fontSize={1} bold>
        {title}
      </Text>
    )}
    {onClose && <StyledCloseButton onClick={onClose} ml="auto" tabIndex="1" />}
  </HeaderWrapper>
)

ModalHeader.displayName = ModalHeader

ModalHeader.propTypes = {
  bg: PropTypes.string,
  onClose: PropTypes.func,
  color: PropTypes.string,
  title: PropTypes.string
}

ModalHeader.defaultProps = {
  bg: 'blue',
  color: 'white'
}

export default ModalHeader
