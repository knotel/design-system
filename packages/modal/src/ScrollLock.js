export default class ScrollLock {
  constructor() {
    this.currentPos = 0
    this.tag = null
    this.head = document.head
  }

  getCurrentPos() {
    return (
      window.pageYOffset ||
      document.documentElement.scrollTop ||
      document.body.scrollTop
    )
  }

  createStyleTag() {
    const css = `
      body {
        overflow: hidden !important;
        position: fixed;
        top: ${this.currentPos}px;
        left: 0;
        width: 100%;
      }
    `
    let tag = document.createElement('style')
    tag.type = 'text/css'
    tag.setAttribute('data-react-scrolllock', '')
    tag.appendChild(document.createTextNode(css))
    this.head.appendChild(tag)
    this.tag = tag
  }

  on() {
    if (!this.head) return
    this.currentPos = this.getCurrentPos()
    this.createStyleTag()
  }

  off() {
    if (!this.tag) return
    this.head.removeChild(this.tag)
    this.tag = null
    window.scrollTo(0, this.currentPos)
  }
}
