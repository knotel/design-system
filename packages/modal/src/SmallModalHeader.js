import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Box } from 'pcln-design-system'

const HeaderBar = styled(Box)`
  height: 8px;
`

const SmallModalHeader = ({ bg }) => <HeaderBar bg={bg} />

SmallModalHeader.displayName = SmallModalHeader

SmallModalHeader.propTypes = {
  bg: PropTypes.string
}

SmallModalHeader.defaultProps = {
  bg: 'green'
}

export default SmallModalHeader
