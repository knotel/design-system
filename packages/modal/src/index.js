export { default as Modal } from './Modal'
export { default as SmallModalHeader } from './SmallModalHeader'
export { default as ModalHeader } from './ModalHeader'
export { default as ScrollLock } from './ScrollLock'
