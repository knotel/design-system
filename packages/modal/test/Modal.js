import React from 'react'
import { render } from 'react-testing-library'
import { Modal } from '../src/index'
import { ThemeProvider, theme } from 'pcln-design-system'

const customRender = (node, ...options) => {
  const rendered = render(
    <ThemeProvider theme={theme}>{node}</ThemeProvider>,
    options
  )

  return {
    ...rendered,
    rerender: (ui, options) =>
      customRender(ui, { container: rendered.container, ...options })
  }
}

describe('Modal', () => {
  const { rerender } = customRender(
    <Modal width={['100px', '200px', '500px']}>
      <div />
    </Modal>
  )
  test('renders when close', () => {
    const { queryByText } = rerender(
      <ThemeProvider>
        <Modal width={['100px', '200px', '500px']}>
          <div data-content>Content</div>
        </Modal>
      </ThemeProvider>
    )
    expect(queryByText('Content')).toBeFalsy()
  })
  test('renders when open', () => {
    const { getByText, container } = rerender(
      <ThemeProvider>
        <Modal width={['100px', '200px', '500px']} isOpen>
          <div data-content>Content</div>
        </Modal>
      </ThemeProvider>
    )
    expect(getByText('Content')).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })
  test('renders when enable overflow and disable close button', () => {
    const { getByText, getByTestId, container } = rerender(
      <ThemeProvider>
        <Modal
          width={['100px', '200px', '500px']}
          isOpen
          enableOverflow
          imgMode
        >
          <div data-content>Content</div>
        </Modal>
      </ThemeProvider>
    )
    expect(getByText('Content')).toBeTruthy()
    expect(getByTestId('pcln-modal-close')).toBeTruthy()
    expect(container.firstChild).toMatchSnapshot()
  })
  test('renders when enable header close button', () => {
    const { getByText } = rerender(
      <ThemeProvider>
        <Modal
          width={['100px', '200px', '500px']}
          isOpen
          header={<div>header</div>}
        >
          <div data-content>Content</div>
        </Modal>
      </ThemeProvider>
    )
    expect(getByText('header')).toBeTruthy()
  })
})
