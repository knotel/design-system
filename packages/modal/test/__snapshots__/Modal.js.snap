// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders when enable overflow and disable close button 1`] = `
.c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
}

.c0 * {
  box-sizing: border-box;
}

<div
  class="c0"
>
  <div
    class="c0"
  />
</div>
`;

exports[`Modal renders when open 1`] = `
.c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
}

.c0 * {
  box-sizing: border-box;
}

<div
  class="c0"
>
  <div
    class="c0"
  />
</div>
`;
